core/src/main/java/org/infinispan/expiration/impl/ClusterExpirationManager.java

```diff
@@ -76,10 +76,14 @@ public void processExpiration() {
                  purgeCandidates.hasNext();) {
                InternalCacheEntry<K, V> e = purgeCandidates.next();
                if (e.canExpire()) {
-                  if (ExpiryHelper.isExpiredMortal(e.getLifespan(), e.getCreated(), currentTimeMillis)) {
-                     handleLifespanExpireEntry(e, true);
-                  } else if (ExpiryHelper.isExpiredTransient(e.getMaxIdle(), e.getLastUsed(), currentTimeMillis)) {
-                     super.handleInMemoryExpiration(e, currentTimeMillis);
+                  // Have to synchronize on the entry to make sure we see the value and metadata at the same time
+                  synchronized (e) {
+                     long lifespan = e.getLifespan();
+                     if (ExpiryHelper.isExpiredMortal(lifespan, e.getCreated(), currentTimeMillis)) {
+                        handleLifespanExpireEntry(e.getKey(), e.getValue(), lifespan, true);
```

Comment:
I am aware. I was being more lazy and to be honest it doesn't save very much as this lock is only on the entry. I can change it up though.